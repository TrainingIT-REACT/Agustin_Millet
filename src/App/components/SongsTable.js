import React, { useState } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import Paper from "@material-ui/core/Paper";
import { Redirect } from "react-router";

const CustomTableCell = withStyles(theme => ({
  head: {
    backgroundColor: "darkblue", //"theme.palette.common.black",
    color: theme.palette.common.white
  },
  body: {
    fontSize: 14
  }
}))(TableCell);

const styles = theme => ({
  root: {
    height: 239,
    width: "100%",
    marginTop: theme.spacing(3),
    overflowX: "auto",
    overflowY: "scroll"
  },
  table: {
    minWidth: "auto"
  },
  row: {
    "&:nth-of-type(odd)": {
      backgroundColor: theme.palette.background.default
    }
  }
});

const parseSeconds = seconds =>
  Number.isInteger(seconds)
    ? `${Math.floor(seconds / 60)}:${seconds % 60}`
    : "";

const SongsTable = props => {
  const { classes, songs } = props;
  const [state, setState] = useState({ pathname: "", redirect: false });

  const handleClick = id => event => {
    setState({ pathname: "/player/" + id, redirect: true });
  };

  return (
    <Paper className={classes.root}>
      {state.redirect === true && <Redirect to={state.pathname} />}
      <Table className={classes.table}>
        <TableHead>
          <TableRow>
            <CustomTableCell>Nombre</CustomTableCell>
            <CustomTableCell align="right">Duraci√≥n</CustomTableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {songs.map(row => (
            <TableRow
              onClick={handleClick(row.id)}
              className={classes.row}
              key={row.id}
            >
              <CustomTableCell component="th" scope="row">
                {row.name}
              </CustomTableCell>
              <CustomTableCell align="right">
                {parseSeconds(row.seconds)}
              </CustomTableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </Paper>
  );
};

SongsTable.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(SongsTable);
