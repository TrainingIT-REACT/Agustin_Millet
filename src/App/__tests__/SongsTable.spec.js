import React from "react";
import { render, mount } from "enzyme";
import SongsTable from "../components/SongsTable";

/**
 * Testing de 2 componentes de presentaciÃ³n
 */

/**
 * SongsTable
 */
describe('SongsTable', () => {
    const songs = {
        items: [
          {
            id: 6,
            name: "Dancing Queen",
            audio: "/music/funky_energy_loop.mp3",
            seconds: 203,
            album_id: 2
          },
          {
            id: 7,
            name: "Fernando",
            audio: "/music/funky_energy_loop.mp3",
            seconds: 203,
            album_id: 2
          },
          {
            id: 8,
            name: "Gimme! Gimme! Gimme! (A Man After Midnight)",
            audio: "/music/funky_energy_loop.mp3",
            seconds: 203,
            album_id: 2
          },
          {
            id: 9,
            name: "Does Your Mother Know",
            audio: "/music/funky_energy_loop.mp3",
            seconds: 203,
            album_id: 2
          },
          {
            id: 10,
            name: "The Name Of The Game",
            audio: "/music/funky_energy_loop.mp3",
            seconds: 203,
            album_id: 2
          }
        ]
      };
      
      const tile = { id: 2 };

/**
 * SongsTable test render con un album de 5 canciones
 */      
  describe("Render", () => {
    let wrapper;
    // Inicializamos el componente en un beforeEach para
    // evitar tener que repetir esta lÃ­nea en cada test
    beforeEach(() => {
      wrapper = render(
        <SongsTable
          songs={songs.items.filter(song => song.album_id === tile.id)}
        />
      );
    });

    it("should add the HTML elements", () => {
      // Comprobamos los distintos aspectos de HTML
      expect(wrapper.is('div[class*="SongsTable-root"]')).toBeTruthy();
      expect(wrapper.find('table').length).toBe(1);
      expect(wrapper.find('table thead').length).toBe(1);
      expect(wrapper.find('tbody tr').length).toBe(5);;
      });
      it('should match the following texts', () => {
      // Comprobamos el texto
      expect(wrapper.find('tbody tr th').first().text()).toBe('Dancing Queen');
      expect(wrapper.find('tbody tr td').first().text()).toBe('3:23');     
    });
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });    
  });

  describe("Shallow", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <SongsTable
          songs={songs.items.filter(song => song.album_id === tile.id)}
        />
      );
    });

    it("should contain the nodes", () => {
        /** 
         * NOTA: Actualmente Enzyme no soporta el acceso a wrapper.state()
         * para componentes funcionales
         */
        // Probar que el nodo padre se ha construido con estilos de MaterialUI
      expect(wrapper.is('WithStyles(SongsTable)')).toBeTruthy();
    });
  });  
});

