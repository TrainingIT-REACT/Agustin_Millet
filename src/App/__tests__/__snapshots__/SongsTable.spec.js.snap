// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongsTable Render should match the snapshot 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded SongsTable-root-1"
>
  <table
    class="MuiTable-root SongsTable-table-2"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-38"
          scope="col"
        >
          Nombre
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-38 MuiTableCell-alignRight"
          scope="col"
        >
          Duraci√≥n
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root SongsTable-row-3"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39"
          scope="row"
        >
          Dancing Queen
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39 MuiTableCell-alignRight"
        >
          3:23
        </td>
      </tr>
      <tr
        class="MuiTableRow-root SongsTable-row-3"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39"
          scope="row"
        >
          Fernando
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39 MuiTableCell-alignRight"
        >
          3:23
        </td>
      </tr>
      <tr
        class="MuiTableRow-root SongsTable-row-3"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39"
          scope="row"
        >
          Gimme! Gimme! Gimme! (A Man After Midnight)
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39 MuiTableCell-alignRight"
        >
          3:23
        </td>
      </tr>
      <tr
        class="MuiTableRow-root SongsTable-row-3"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39"
          scope="row"
        >
          Does Your Mother Know
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39 MuiTableCell-alignRight"
        >
          3:23
        </td>
      </tr>
      <tr
        class="MuiTableRow-root SongsTable-row-3"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39"
          scope="row"
        >
          The Name Of The Game
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-body-39 MuiTableCell-alignRight"
        >
          3:23
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
